### PR TITLE
Add grip-mode to markdown layer from github layer

### PR DESCRIPTION
**Pull request by robbyoconnor**
Since the github layer was deprecated in
cb786cc0711e0b23790185ef79012812dcfbe768, add it to the markdown layer.

### CODE REVIEW

**Comment by bryce-carson** [comment]
Thank you, @robbyoconnor for the pull request. This looks fine, but I did create some review comments.

I believe this should be placed within the Markdown layer, however, as the functionality provided by Grip is related to GitHub and its specific rendering of (primarily) Markdown in `README.md` files (and optionally `README.org` files). Do you agree, @lebensterben, that this should be added to the Markdown layer instead?

**Comment by robbyoconnor** [comment]
> I believe this should be placed within the Markdown layer, however, as the functionality provided by Grip is related to GitHub and its specific rendering of (primarily) Markdown in `README.md` files (and optionally `README.org` files). Do you agree, @lebensterben, that this should be added to the Markdown layer instead?

No, because this is specific to GitHub technically, which goes with git since GitHub is a forge.

**Comment by bryce-carson** [comment]
> > I believe this should be placed within the Markdown layer, however, as the functionality provided by Grip is related to GitHub and its specific rendering of (primarily) Markdown in `README.md` files (and optionally `README.org` files). Do you agree, @lebensterben, that this should be added to the Markdown layer instead?
> 
> No, because this is specific to GitHub technically, which goes with git since GitHub is a forge.

GitHub _is a forge_, yes. The Git layer adds functionality to work with _any_ forge, not just Github.

The Grip Python package provides functionality specific to GitHub's rendering of (primarily) Markdown, as I already said. If you `C^f` the devel version of the Git layer README you'll notice it makes no reference to markdown, and the only reference to Org is the `orgit` package included in Magit.

In short, it's not related to the `git` command line tool (Git), version control *systems* (SVN, Bazaar, Git, etc.) and should belong to the markdown layer because that is what it is related to.

GitHub != Git layer.

**Comment by robbyoconnor** [comment]
I'll move it.

**Comment by robbyoconnor** [comment]
@bryce-carson moved.

File: layers/+lang/markdown/README.org
Comment (robbyoconnor) [inline layers/+lang/markdown/README.org:169]: Not sure about this keybinding

**Comment by bryce-carson** [comment]
Once a maintainer responds to the comments I made regarding the keybinding for this, it'll probably be ready to merge.

You can always keep `g p` for your personal config if you prefer. :)

File: layers/+lang/markdown/packages.el
Comment (robbyoconnor) [inline layers/+lang/markdown/packages.el:68]: Not sure about this keybinding

**Comment by bryce-carson** [comment]
The `safe_join` function in Flask [was deprecated recently](https://github.com/pallets/flask/blob/b109b092a95036de875246f05cdc377a2c7faab8/CHANGES.rst#id5), so the current version of Grip on PyPI is broken (joeyespo/grip@be091045a1c1175850c766b9793edbfddca88afd).

This cannot be merged until it is fixed, as users without existing, outdated Flask installations will find that the package is broken.

**Comment by robbyoconnor** [comment]
@bryce-carson was it deprecated or removed because deprecated doesn't equate to removal.

**Comment by bryce-carson** [comment]
> @bryce-carson was it deprecated or removed because deprecated doesn't equate to removal.

It was removed, pardon me. Previously it *was* deprecated. Now it is gone and functionality in Grip is broken.

**Comment by robbyoconnor** [comment]
Yeah the words matter a lot here! Let's hold off but this can be safely merged... it's functionality that exists already

**Comment by bryce-carson** [comment]
> Yeah the words matter a lot here! Let's hold off but this can be safely merged... it's functionality that exists already

True. Anyone trying the currently deprecated GitHub layer, without already having Flask as a requirement on their system, will find it's broken regardless.

The merge would be good with the suggested keybinding and the fixed init function, @robbyoconnor.

File: layers/+lang/markdown/README.org
Comment (lebensterben) [inline layers/+lang/markdown/README.org:56]: No need to include installation instruction for specific distributuon.

File: layers/+lang/markdown/README.org
Comment (lebensterben) [inline layers/+lang/markdown/README.org:168]: I don't think this is necessary. Let user call grip-mode via M-x.

**Review by lebensterben — CHANGES_REQUESTED** [review]

**Comment by lebensterben** [comment]
I never thought it's necessary to preview any markup language...

But since some people want it we can add it, but preferably behind a layer config variable.

File: layers/+lang/markdown/README.org
Comment (robbyoconnor) [reply layers/+lang/markdown/README.org:56]: This is literally copy/paste from the `github` layer.

File: layers/+lang/markdown/README.org
Comment (robbyoconnor) [reply layers/+lang/markdown/README.org:168]: I do -- I don't wanna do it -- why not have a leader key

**Comment by lebensterben** [comment]
also, this should be in the markdown layer. because GitHub flavored markdown is widely used, not limiting to GitHub.

**Comment by robbyoconnor** [comment]
@lebensterben 

> also, this should be in the markdown layer. because GitHub flavored markdown is widely used, not limiting to GitHub.

Check again. It is.

File: layers/+lang/markdown/README.org
Comment (bryce-carson) [reply layers/+lang/markdown/README.org:56]: I proposed changes earlier. We may as well rewrite the description while we're moving it from the deprecated layer.

**Comment by bryce-carson** [comment]
@robbyoconnor @lebensterben I didn't realize I had to "finish" a review. I wrote those review comments many hours ago.

They're there and availble now, I guess.

File: layers/+lang/markdown/README.org
Comment (lebensterben) [reply layers/+lang/markdown/README.org:168]: SPC leader key is too precious.

File: layers/+source-control/git/README.org
Comment (lebensterben) [reply layers/+source-control/git/README.org]: yes. we should not include any ditro specific instruction.

**Comment by lebensterben** [comment]
on the installation command,

for technical reasons it's prefered to use `python3 -m pip install` in a *certain environment*.

obviously this easily gets very complicated due to virtual environment...

So either just tell the user to use `pip install` or just put the link there and let them figure out themselves...

**Comment by lebensterben** [comment]
if you really want a key binding, just make a major mode key binding for markdown-mode (which is prefixed with SPC m)

**Comment by robbyoconnor** [comment]
Will do @lebensterben

File: layers/+lang/markdown/README.org
Comment (dalanicolai) [reply layers/+lang/markdown/README.org:168]: I would prefer to 'more actively' inform users about this functionality by adding a keybinding for it. But why not leave it just under "ghp" as "gh" already exists and grip is 'more or less' github specific (users can still also call via `M-x` of course)

**Comment by dalanicolai** [comment]
Actually the main reason for adding grip-mode, beside the already, in the markdown layer available `vmd-mode` that also does github markdown rendering, is that grip-mode supports previewing `org` files directly. Therefore, I would still vote for a keybinding under `SPC g h` (although I would not mind another major mode keybinding in markdown mode).

But like @robbyoconnor my idea is that this package is more github related than markdown related (and therefore I did not protest when putting it in the git layer, despite the git layer not being a github layer specifically. Because then we would not be required to install the markdown layer to preview `org` files)... Anyway, in the end, I'll probably support any choice you make.

File: layers/+lang/markdown/README.org
Comment (lebensterben) [reply layers/+lang/markdown/README.org:168]: @dalanicolai 

nope it's not GitHub specific.
GFM is widely used and is one of top makrdown implementations.

though grip makes it sounds like it's used for GitHub only, this package can be used for previewing all GFM.

File: layers/+lang/markdown/README.org
Comment (robbyoconnor) [reply layers/+lang/markdown/README.org]: Resolved.

File: layers/+lang/markdown/packages.el
Comment (robbyoconnor) [reply layers/+lang/markdown/packages.el:68]: @bryce-carson what does the T stand for?

**Comment by robbyoconnor** [comment]
Okay, so I thought about this. It's also relevant to Org Mode, so what do we do about that?

File: layers/+lang/markdown/packages.el
Comment (robbyoconnor) [reply layers/+lang/markdown/packages.el:64]: Nuked the keybinding.

File: layers/+lang/markdown/README.org
Comment (robbyoconnor) [inline layers/+lang/markdown/README.org:165]: This should work.

**Comment by winsphinx** [comment]
What difference is in previewing between vmd-mode and grip-mode?

**Comment by lebensterben** [comment]
they use different program, vmd vs grip. some people have preference of one of NodeJS and Python over the other.

**Comment by lebensterben** [comment]
vmd only supports GitHub access token from command line (and yes this is not even documented).

grip is better and you can specify that in a config file or let it read the password from a password manager.

So grip is better in this regard.

---

Both support auto refreshing so that gives you a live preview similar to other markdown editors. BUT, without access token or password, you'd be rate-limited very soon.

**Comment by robbyoconnor** [comment]
I wouldn't be offended if this is rejected but how with the org layer?

**Comment by lebensterben** [comment]
@robbyoconnor 

well... if you want, move it to a tool/grip layer...

**Comment by robbyoconnor** [comment]
@lebensterben we could make a layer for preview tools as well

**Comment by lebensterben** [comment]
that's a good idea actually.

**Comment by bryce-carson** [comment]
The Org functionality is a lesser feature, no?

Could be easily solved by using a delist + set leader keys for major mode to provide a toggle, or better use the toggle functions.

**Comment by robbyoconnor** [comment]
@bryce-carson The idea of moving them to their own layer is useful. What if they don't have the markdown layer?

**Comment by bryce-carson** [comment]
> @bryce-carson The idea of moving them to their own layer is useful. What if they don't have the markdown layer but the org layer?

I'm not sure about that case.

A layer for previewers may be useful, but I don't know too much about how multiple layers work together.

There's also the point that if you wanted just markdown and markdown previewing, you would be getting a lot of other previewers that you're not using.

What's best for the direction to take applications like this in Spacemacs is hard to say. I don't know.

**Comment by lebensterben** [comment]
just make a tools/previewer layer, with vmd-mode and grip-mode. (probably other packages)

each package is gated behind a layer config variable.

**Comment by robbyoconnor** [comment]
I'll do it this weekend. Going to close this and make a new PR

**Pull request closed without merging**
